gen/perunDataGenerator.pm

```diff
@@ -18,18 +18,58 @@ sub generateUsersDataInJSON {
 	perunServicesInit::init;
 
 	my $DIRECTORY = perunServicesInit::getDirectory;
-	my $data = perunServicesInit::getFlatData;
+	my $data = perunServicesInit::getHashedHierarchicalData;
+	my $agent = perunServicesInit->getAgent;
+	my $attributesAgent = $agent->getAttributesAgent;
+	my $servicesAgent = $agent->getServicesAgent;
+	my $service = $servicesAgent->getServiceByName( name => $::SERVICE_NAME);
+
+	my @requiredAttributesDefinitions = $attributesAgent->getRequiredAttributesDefinition(service => $service->getId);
+	my @userRequiredAttributes = ();
+	my @userFacilityRequiredAttributes = ();
+	foreach my $attrDef (@requiredAttributesDefinitions) {
+		# if attribute's namespace starts with "urn:perun:user:"
+		my $o = index $attrDef->getNamespace, "urn:perun:user:";
+		if ($o == 0) {
+			push @userRequiredAttributes, $attrDef;
+			next;
+		}
+		$o = index $attrDef->getNamespace, "urn:perun:user_facility:";
+		if ($o == 0) {
+			push @userFacilityRequiredAttributes, $attrDef;
+			next;
+		}
+	}
 	my @users;
 
 	####### prepare data ######################
-	foreach my $user (($data->getChildElements)[1]->getChildElements) {
+	my %usersIds = ();
+	foreach my $memberId ($data->getMemberIdsForFacility()) {
+		my $userId = $data->getUserIdForMember(member => $memberId);
+		if (exists($usersIds{$userId})) {
+			next;
+		} else {
+			$usersIds{$userId} = 0;
+		}
 		my $userData = {};
-		foreach my $uAttribute ($user->getAttributes) {
+
+		foreach my $userAttribute (@userRequiredAttributes) {
+			my $attrValue = $data->getUserAttributeValue(member => $memberId, attrName => $userAttribute->getName);
+			# In case there is an undefined boolean attribute, we have to change it to false
+			if ($userAttribute->getType eq "boolean" && !defined $attrValue) {
+				$userData->{$userAttribute->getName} = \0;
+			} else {
+				$userData->{$userAttribute->getName} = $attrValue;
+			}
+		}
+
+		foreach my $userFacilityAttribute (@userFacilityRequiredAttributes) {
+			my $attrValue = $data->getUserFacilityAttributeValue(member => $memberId, attrName => $userFacilityAttribute->getName);
 			# In case there is an undefined boolean attribute, we have to change it to false
-			if ($uAttribute->getType eq "boolean" && !defined $uAttribute->getValue) {
-				$userData->{$uAttribute->getName} = \0;
+			if ($userFacilityAttribute->getType eq "boolean" && !defined $attrValue) {
+				$userData->{$userFacilityAttribute->getName} = \0;
 			} else {
-				$userData->{$uAttribute->getName} = $uAttribute->getValue;
+				$userData->{$userFacilityAttribute->getName} = $attrValue;
 			}
 		}
 		push @users, $userData;
```

Comment:
You can use `canonical` when printing JSON to get better / comparable outputs.
```perl
print FILE JSON::XS->new->utf8->pretty->canonical->encode(\@users);
```

Reply:
Added.